packages/plugin-react/src/index.js

```diff
@@ -1,235 +1,210 @@
 // @ts-check
-const fs = require('fs')
-const { transformSync, ParserOptions } = require('@babel/core')
-const runtimePublicPath = '/@react-refresh'
-const runtimeFilePath = require.resolve(
-  'react-refresh/cjs/react-refresh-runtime.development.js'
-)
-
-const runtimeCode = `
-const exports = {}
-${fs.readFileSync(runtimeFilePath, 'utf-8')}
-function debounce(fn, delay) {
-  let handle
-  return () => {
-    clearTimeout(handle)
-    handle = setTimeout(fn, delay)
-  }
-}
-exports.performReactRefresh = debounce(exports.performReactRefresh, 16)
-export default exports
-`
-
-const preambleCode = `
-import RefreshRuntime from "__BASE__${runtimePublicPath.slice(1)}"
-RefreshRuntime.injectIntoGlobalHook(window)
-window.$RefreshReg$ = () => {}
-window.$RefreshSig$ = () => (type) => type
-window.__vite_plugin_react_preamble_installed__ = true
-`
+import babel from '@babel/core'
+import viteReactJsx, { babelImportToRequire } from 'vite-react-jsx'
+import {
+  addRefreshWrapper,
+  isRefreshBoundary,
+  preambleCode,
+  runtimeCode,
+  runtimePublicPath
+} from './fast-refresh.js'
 
 /**
- * Transform plugin for transforming and injecting per-file refresh code.
- *
- * @type {import('.').default}
+ * @param {import('./').Options} opts
  */
-function reactRefreshPlugin(opts) {
-  let shouldSkip = false
+export default function viteReact(opts = {}) {
   let base = '/'
+  let projectRoot = process.cwd()
+  let isProduction = true
+  let skipFastRefresh = opts.fastRefresh === false
 
-  return {
-    name: 'react-refresh',
+  const userPlugins = opts.babel?.plugins || []
+  const userParserPlugins =
+    opts.parserPlugins || opts.babel?.parserOpts?.plugins || []
 
+  /** @type {import('vite').Plugin} */
+  const viteBabel = {
+    name: 'vite:babel',
     enforce: 'pre',
+    async transform(code, id, ssr) {
+      if (/\.[tj]sx?$/.test(id)) {
+        const plugins = [...userPlugins]
+
+        /** @type {typeof userParserPlugins} */
+        const parserPlugins = [
+          ...userParserPlugins,
+          'jsx',
+          'importMeta',
+          // This plugin is applied before esbuild transforms the code,
+          // so we need to enable some stage 3 syntax that is supported in
+          // TypeScript and some environments already.
+          'topLevelAwait',
+          'classProperties',
+          'classPrivateProperties',
+          'classPrivateMethods',
+          'objectRestSpread'
```

Comment:
Oops, this parser plugin can be removed, as it's enabled by default.

https://github.com/babel/babel/blob/813f2a3cce04a5366de7c1cc370652db64507305/packages/babel-plugin-syntax-typescript/src/index.js#L36-L38